File: js/algorithm/radix_sort.js
/**
 * 使用队列来实现基数排序 radix sort
 * 一种无需比较和交换的算法
 * https://baike.baidu.com/item/%E5%9F%BA%E6%95%B0%E6%8E%92%E5%BA%8F/7875498?fr=aladdin
 */

const Queue = require('../data_structures/queue');

function radixSort(args = []) {
    // 需要创建10个队列 bin0 - bin9;
    const binArr = []; // 用于存储binx的对象
    let dataArr = [];

    for (let i = 0; i <= 9; i++) {
        binArr.push({ id: `bin${i}`, queue: new Queue() });
    }

    // 对数组中的数取个位数入队, 进行第一次排序;
    for (let i = 0; i < args.length; i++) {
        const item = args[i];
        const itemSingle = item % 10;
        binArr[itemSingle].queue.put(item)
    }
    // 第一次排序后依次出队，得到第一次排序的结果
    dataArr = popArr(binArr);

    // 对弹出的数取十位数再排序一次
    for (let i = 0; i < dataArr.length; i++) {
        const item = dataArr[i];
        const itemSingle = Math.floor(item / 10);
        binArr[itemSingle].queue.put(item)
    }

    // 第二次排序后，返回最终结果
    return popArr(binArr)

}

// 弹出bin队列
function popArr(queueArr) {
    const singleArr = [];
    for (let i = 0; i < queueArr.length; i++) {
        const curQue = queueArr[i].queue;
        while(curQue.size() > 0) {
            singleArr.push(curQue.get());
        }
    }
    return singleArr;
}

module.exports = radixSort

File: py/algorithm/radix_sort.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

# 定义一个模块需要上面的声明， 这样才可以被其他模块执行

# 使用队列实现基数排序 py

from queue import Queue
from math import floor


def radix(args=[]):
    # 创建10个队列 bin0 - bin10
    bin_arr = []
    data_arr = []
    for i in range(10):
        bin_item = {
            'id': 'bin%s' % i,
            'queue': Queue()
        }
        bin_arr.append(bin_item)
    # 对数组中对个位数进行一次入队排序
    for item in args:
        cur_item = item % 10 # 取余数
        cur_arr = bin_arr[cur_item] # bin0, bin1
        cur_arr['queue'].put(item)
    # 第一次排序后依次出队，得到第一次排序的结果
    data_arr = pop_arr(bin_arr)

    # 对数组对十位数再进行一次入队排序
    for item in data_arr:
        cur_item = floor(item / 10) # 取十位数
        cur_arr = bin_arr[cur_item] # bin0, bin1
        cur_arr['queue'].put(item)
    
    data_arr = pop_arr(bin_arr)

    return data_arr


def pop_arr(bin_arr=[]):
    single_arr = []
    for bin_x in bin_arr:
        while bin_x['queue'].qsize():
            single_arr.append(bin_x['queue'].get())
    return single_arr
